File: benchmarks/live_suite/matmul.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>

int64_t run_matmul(int n) {
    int64_t *a = (int64_t *)malloc(n * n * sizeof(int64_t));
    int64_t *b = (int64_t *)malloc(n * n * sizeof(int64_t));
    int64_t *c = (int64_t *)calloc(n * n, sizeof(int64_t));

    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            a[i * n + j] = (i + j) % 100;
            b[i * n + j] = (i * j) % 100;
        }
    }

    for (int i = 0; i < n; i++) {
        for (int k = 0; k < n; k++) {
            int64_t a_ik = a[i * n + k];
            for (int j = 0; j < n; j++) {
                c[i * n + j] += a_ik * b[k * n + j];
            }
        }
    }

    int64_t res = c[n * n - 1];
    free(a); free(b); free(c);
    return res;
}

int main(void) {
    int64_t checksum = 0;
    for (int iter = 0; iter < 10; iter++) {
        checksum = run_matmul(128);
    }
    printf("Matmul Checksum: %lld\n", (long long)checksum);
    return 0;
}

File: benchmarks/live_suite/matmul.rs
fn run_matmul(n: usize) -> i64 {
    let mut a = vec![0i64; n * n];
    let mut b = vec![0i64; n * n];
    let mut c = vec![0i64; n * n];

    for i in 0..n {
        for j in 0..n {
            a[i * n + j] = ((i + j) % 100) as i64;
            b[i * n + j] = ((i * j) % 100) as i64;
        }
    }

    for i in 0..n {
        for k in 0..n {
            let a_ik = a[i * n + k];
            for j in 0..n {
                c[i * n + j] += a_ik * b[k * n + j];
            }
        }
    }

    c[n * n - 1]
}

fn main() {
    let mut checksum = 0;
    for _ in 0..10 {
        checksum = run_matmul(128);
    }
    println!("Matmul Checksum: {}", checksum);
}
